1.4 *FINAL*

Voidless7125 opened the pull request:
Build.ps1: Added auto installing of  `VSSetup`.
sandbox_run.ps1: Fixed having two shortcuts of LockDown Browser past 2.1.3.00. 
GetSystemMetrics-Hook.cpp: Massive changes, disables checking for alt-tab or minimizing altogether. (Fixing Win 10 issues)

Still working on it.

gucci-on-fleek commented on src/pch.h:
Using precompiled headers just seems like unnecessary complexity to me, unless the speed boost really is quite significant.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -19,23 +69,457 @@ int WINAPI Hooked_GetSystemMetrics(int nIndex)
+HANDLE WINAPI MySetClipboardData(UINT uFormat, HANDLE hMem)
+{
+    // This custom function does nothing
+    std::cout << "SetClipboardData hook called, but not setting clipboard data." << std::endl;
```
I'm mostly against logging inside the hooked functions since it's really not very useful to know when a hooked function was called outside of debugging, and doing pretty much anything at all inside of a hook is somewhat risky. I wouldn't be opposed to keeping this inside of an 
```c
#ifdef DEBUG
std::cout << "stuff"
#endif
```
guard though.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -19,23 +69,457 @@ int WINAPI Hooked_GetSystemMetrics(int nIndex)
+
+void InstallHook()
+{
+    std::cout << "Installing hooks..." << std::endl;
```
Are you certain that iostreams doesn't allocate dynamic/heap memory? (I honestly have no idea) Because it's not safe to allocate any memory from `DllMain`.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
Sorry, but I'm mostly against the changes in this file, but as always, you can try and convince me otherwise. A few comments:

1. I'd really like to keep the number of hooked functions to an absolute minimum because:

    a. Detection surface area. This one is fairly unimportant since there are already a dozen ways that Respondus could detect us if they were trying, but something that's generally good to keep in mind.

    b. Code complexity. I liked that the original repository really only had a hundred lines of code in it since it meant that any users wouldn't really have to trust us—the code was short enough that you could “easily” read and understand most of it in an hour. Now, the PowerShell files have somewhat moved away from this, but most of those changes were in response to real issues.

    c. Risk. Hooking functions is fairly risky, and it's in fact uniquely risky, since unlike the other interventions, it could easily crash the _Lockdown Browser_ in the middle of an exam, or even worse, it could corrupt some internal state and lead to unpredictable behaviour.

    d. Intent. The purpose of this repository is fairly well summed up in [`Readme.md` § Purpose](https://github.com/gucci-on-fleek/lockdown-browser?tab=readme-ov-file#purpose) and your last paragraph in https://github.com/gucci-on-fleek/lockdown-browser/issues/121#issue-2836250499. Now, I know that most users are probably using this for cheating, but I'd still rather keep this “sandboxing-oriented” rather than “cheating-oriented”, and some of the new hooked functions seem to push this a little more towards the later.

Again, sorry for the rough feedback, but I'd rather get this out here before you put too much work into things. I definitely appreciate the effort though. (And again, I can maybe be convinced on some of these points).

Voidless7125 replied:
I've removed the changes. I'm marking all this as resolved because it's gone. You had good points, and don't worry about being harsh; all is good. =)

gucci-on-fleek commented on runtime_directory/sandbox_run.ps1:
```diff
@@ -80,13 +80,18 @@ function Install-LockdownBrowser {
-        Write-Log "Removed existing LockDown Browser shortcut from public desktop."
+        $shortcut = "C:\Users\Public\Desktop\LockDown Browser.lnk"
+        1..20 | ForEach-Object {
+            if (Test-Path $shortcut) { return }
```
Checking to see if a file exists is fairly cheap, so you might as well reduce the sleep delay to speed up the installation by a little bit.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -19,23 +69,457 @@ int WINAPI Hooked_GetSystemMetrics(int nIndex)
+}
 
-BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+void InstallFocus()
```
I think that I've pointed out most of the problematic code below, but here's some general reading about what is and isn't safe to do inside of `DllMain`:

- https://www.google.com/search?q=site%3Adevblogs.microsoft.com%2Foldnewthing+dllmain
- https://learn.microsoft.com/en-us/windows/win32/dlls/dllmain
- https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-best-practices#general-best-practices

gucci-on-fleek commented on build.ps1:
```diff
@@ -9,6 +9,7 @@ Set-StrictMode -Version 3
 Set-Location $PSScriptRoot
 
 # Import the VSSetup module to use Get-VSSetupInstance function
+Install-Module VSSetup -Scope CurrentUser
 Import-Module VSSetup
```
This would maybe be better inside of `initialize_vs`?

Voidless7125 replied:
Noted.

Voidless7125 replied:
Finished.

gucci-on-fleek reviewed (commented):
I know that you're still working on the code, but I figured that I'd send in a few comments before you put too much work into anything. The PowerShell changes look good to me aside from a few nits, but I'm fairly concerned about the C++ changes—I appreciate the effort, but I'm not entirely sure that all of the changes are suited for this project.

Voidless7125 commented:
@gucci-on-fleek, I read all the comments, and no worries. I was just integrating all of it into the file. I didn't change much other than fixing compile errors.
Most of the C++ changes are oriented for cheating; I will eliminate the keyboard shortcut changes and the clipboard hack; I only need the window detection disabled.
And yeah, all the unsafe code will be fixed. I may move this out of scope and launch 1.4 without these changes.

Voidless7125 commented on build.ps1:
```diff
@@ -25,9 +26,16 @@ function Write-Log {
-        Write-Log "Error: No Visual Studio Build Tools instances found"
-        throw "No Visual Studio instances found"
+    if (-not $vs_instances.Count -eq 0) {
+        $answer = Read-Host "No Visual Studio Build Tools instances found. This can sometimes be wrong. If you have installed this press Y. (Y/N)"
```
Fixed.

Voidless7125 commented:
Okay I got one last thing to do and this will be ready.

Voidless7125 commented:
Whoops, forgot to push a change. Hang tight.

Voidless7125 commented:
Now it is ready.

Voidless7125 commented:
Hang tight; I just broke the Powershell file. I was working on it and pushed it to the wrong place... :P

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -6,71 +6,118 @@
 {
-    /* This is all pretty much copied from the Detours docs. I can't really pretend to understand most of it */
-    // I also can't pretend I understand this, but hey error checking would be nice - Voidless
-    (void)hinstDLL; // Discard the unused parameters
-    (void)lpvReserved;
+    return TRUE; // Simulate success without terminating
+}
 
-    if (DetourIsHelperProcess())
+VOID WINAPI MyExitProcess(UINT uExitCode)
+{
+    // Do nothing
+}
```
What do these two functions do? I mean, yes, the block the _Lockdown Browser_ from closing processes, but why is this needed? Which processes does this prevent it from closing? Won't it notice that it failed to close these processes when it calls `EnumProcesses` (which is non-trivial to work around [since we can't easily hook `EnumProcesses`](https://github.com/gucci-on-fleek/lockdown-browser/issues/4#issuecomment-1159352265)).

To be clear, I'm not opposed to adding these hooks, I just want to make sure that I know what they do.

Voidless7125 replied:
"Which processes does this prevent it from closing?" Anything. So, any app in the VM that may get blocked in a later update won't close. (Say another VM compute exe gets blocked; you can click close, and it will do nothing.)

"Won't it notice that it failed to close these processes when it calls EnumProcesses (which is non-trivial to work around https://github.com/gucci-on-fleek/lockdown-browser/issues/4#issuecomment-1159352265)." Actually, no, as the lockdown browser assumes it closed without checking. If they add a check later on, we would *100%* get flagged. I might be able to make some simple code that reports that banned exes are "not there" when they are.

Voidless7125 replied:
That totally makes sense—detection states (4) and (5) should be avoided at all costs because they trigger an issue beyond a simple message. On the other hand, detection state (3) is relatively harmless because it happens before login, meaning there’s no real penalty.

Given that both (B) and (D) are common, the real question is whether the risk of moving from (3) → (2) (which is fine) outweighs the risk of moving from (3) → (5) (which is catastrophic). Since (5) has long-term consequences and (3) is essentially a free pass, your caution makes great sense.

The safer approach is **not** to hook the functions, as the worst-case scenario (going from (3) to (5)) is too severe to justify the potential benefit of avoiding (3) altogether. The patch provides a short-term convenience but opens the door to a much riskier outcome.

That being said, If I can patch it so that the lockdown browser never detects the blocked software in the first place, then that would theoretically move you from state (3) → state (1)—no detection at all. That’s the ideal scenario because it eliminates the entire risk of escalation (no warnings, no bans, no external consequences).

The main questions then are:

Lockdown browser has multiple detection methods (e.g., scanning process lists, checking window titles, querying system APIs, etc.), and I must block and find all of them. I can use the help of the repo I used last time, but it got patched for 2.1.3.00. There may be more. 

Also, if the browser gets an update with improved detection, will my patch still work or trigger a worse outcome (e.g., detecting tampering)?

Also, If an update makes it verifies its code or behavior (e.g., using checksums, anti-debugging techniques, etc.), then modifying its ability to detect software might trigger state (5) directly.
If it logs inconsistencies and reports them later, you could end up with a delayed ban. (Although this has never happened, it doesn't mean it can't)

If I apply this patch and it works now, what happens when Respondus releases an update?
If they improve detection, would this immediately escalate to state (4) or (5)?
If we can 100% confirm that the patch completely blocks detection without triggering additional scrutiny, then it’s a clear improvement over both (3) and (2). However, if there’s any chance that it raises suspicion or causes unintended side effects, then you'd be in a worse situation.

It comes down to whether the risk of Respondus detecting the patch outweighs the benefit of avoiding detection altogether. If I (and you) are confident in preventing detection without triggering integrity checks, then having the patch would be the safest approach.

Sorry this is long, but I need to cover all bases here.

Voidless7125 replied:
Also, as for detection of injecting, here's what they could do. (I think, if this isn't valid, let me know)

1. DLL Enumeration (Process Scanning)
The lockdown browser can list all loaded DLLs in its process using:
EnumProcessModules()
NtQueryInformationProcess()
CreateToolhelp32Snapshot()
If it checks for unexpected DLLs, yours could be flagged unless you hide it.

Bypassing This:

Manual mapping (load the DLL in memory without actually "loading" it as a module).
Unlinking from PEB (removes the DLL from Windows' internal module list).

2. Code-Signing & Integrity Checks
If the lockdown browser verifies that only signed or whitelisted DLLs are loaded, an unsigned DLL could be flagged.
Some software checks for newly attached DLLs by re-scanning modules during runtime.
Bypassing This:

If the lockdown browser only checks itself at startup, injecting after it loads could avoid detection.
If it only checks the .exe signature and not modules, your DLL might not be an issue.

3. Anti-Debugging & Memory Scans
Some software detects unusual memory changes or injected threads.
That might be detectable if your DLL creates a new thread (e.g., CreateRemoteThread()).

Bypassing This:

Use APC injection or thread hijacking instead of CreateRemoteThread().
Hook detection functions like EnumProcessModules() and make them "ignore" your DLL.

4. Behavior-Based Detection
Even if the DLL itself isn’t flagged, what it does might be.
The lockdown browser might catch on if it blocks process enumeration, modifies detection functions, or hooks system calls.

 Bypassing This:
Instead of directly blocking detection, consider returning fake but "safe" results (e.g., returning a normal-looking process list or not including blocked items but allowing for allowed items)

Conclusion
A simple injection is probably safe if the lockdown browser doesn’t scan for injected DLLs. (Currently, it doesn't?)
If it does check for unexpected modules, then manual mapping or unlinking from PEB would be needed.
The safest bet is to inject after startup, avoid modifying critical structures, and ensure any function hooking is stealthy.

gucci-on-fleek reviewed (commented):
1. Still working on various homework assignments, so this review is brief/incomplete
2. I'm fairly sleep deprived, so some of my comments are probably wildly inaccurate. Ignore as necessary :)

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,68 @@
+    wtite_log "Zipping logs folder as requested by -l flag"
+    $zipFile = Join-Path $PSScriptRoot "logs.zip"
+    if (Test-Path $zipFile) { Remove-Item $zipFile -Force }
+    Compress-Archive -Path "./logs/*" -DestinationPath $zipFile
```
Sounds good. I'll start working on that.

Voidless7125 replied:
Done.

Voidless7125 replied:
Here's what it looks like now:

```log
=== Build.log ===

2025-02-14 22:50:15 - Removal complete
2025-02-14 22:50:15 - ----------------------------------------
2025-02-14 22:50:15 - Build script started
2025-02-14 22:50:15 - Collecting system information...
2025-02-14 23:26:41 - ----------------------------------------
2025-02-14 23:26:41 - Build script started
2025-02-14 23:26:41 - Collecting system information...
2025-02-14 23:26:50 - Logs being put in one file as requested by -logs flag


=== sandbox_run.log ===

2025-02-14 23:15:00 - ----------------------------------------
2025-02-14 23:15:00 - Script started.
2025-02-14 23:15:00 - Removing system information...
2025-02-14 23:15:05 - Removed system information successfully.
2025-02-14 23:15:09 - Extracting Lockdown Browser...
2025-02-14 23:15:49 - Installing Lockdown Browser...
2025-02-14 23:15:49 - Setup executed.
2025-02-14 23:15:53 - Removed existing LockDown Browser shortcut.
2025-02-14 23:15:53 - Registering URL protocol(s)...
2025-02-14 23:15:53 - Successfully set item property for URL protocol rldb.
2025-02-14 23:15:53 - Creating run script on desktop...
2025-02-14 23:15:54 - Removed existing LockDown Browser start menu shortcut.
2025-02-14 23:15:54 - Run script, shortcut, and start menu shortcut created.
2025-02-14 23:16:03 - Script completed.


=== setup.log ===

[InstallShield Silent]
Version=v7.00
File=Log File
[ResponseResult]
ResultCode=0
[Application]
Name=Respondus LockDown Browser 2
Version=2.00.0000
Company=Respondus
Lang=0409
```

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,68 @@
 # SPDX-License-Identifier: MPL-2.0+
 # SPDX-FileCopyrightText: 2020-2025 gucci-on-fleek and Voidless7125
 
+param(
```
Good Idea. And the readme also has to be updated with the new ones, I won't forget.

Voidless7125 replied:
And done.

Voidless7125 replied:
Here's what it looks like now:

```log
=== Build.log ===

2025-02-14 22:50:15 - Removal complete
2025-02-14 22:50:15 - ----------------------------------------
2025-02-14 22:50:15 - Build script started
2025-02-14 22:50:15 - Collecting system information...
2025-02-14 23:26:41 - ----------------------------------------
2025-02-14 23:26:41 - Build script started
2025-02-14 23:26:41 - Collecting system information...
2025-02-14 23:26:50 - Logs being put in one file as requested by -logs flag


=== sandbox_run.log ===

2025-02-14 23:15:00 - ----------------------------------------
2025-02-14 23:15:00 - Script started.
2025-02-14 23:15:00 - Removing system information...
2025-02-14 23:15:05 - Removed system information successfully.
2025-02-14 23:15:09 - Extracting Lockdown Browser...
2025-02-14 23:15:49 - Installing Lockdown Browser...
2025-02-14 23:15:49 - Setup executed.
2025-02-14 23:15:53 - Removed existing LockDown Browser shortcut.
2025-02-14 23:15:53 - Registering URL protocol(s)...
2025-02-14 23:15:53 - Successfully set item property for URL protocol rldb.
2025-02-14 23:15:53 - Creating run script on desktop...
2025-02-14 23:15:54 - Removed existing LockDown Browser start menu shortcut.
2025-02-14 23:15:54 - Run script, shortcut, and start menu shortcut created.
2025-02-14 23:16:03 - Script completed.


=== setup.log ===

[InstallShield Silent]
Version=v7.00
File=Log File
[ResponseResult]
ResultCode=0
[Application]
Name=Respondus LockDown Browser 2
Version=2.00.0000
Company=Respondus
Lang=0409
```

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,68 @@
+# If removal flag -r is passed, delete specified directories and files
+if ($r) {
+    wtite_log "Removal flag specified (-r): Cleaning directories and files"
+    "1" | git clean -ixd
```
Also a good idea. Will do that.

Voidless7125 replied:
Done.

Voidless7125 replied:
Here's what it looks like now:

```log
=== Build.log ===

2025-02-14 22:50:15 - Removal complete
2025-02-14 22:50:15 - ----------------------------------------
2025-02-14 22:50:15 - Build script started
2025-02-14 22:50:15 - Collecting system information...
2025-02-14 23:26:41 - ----------------------------------------
2025-02-14 23:26:41 - Build script started
2025-02-14 23:26:41 - Collecting system information...
2025-02-14 23:26:50 - Logs being put in one file as requested by -logs flag


=== sandbox_run.log ===

2025-02-14 23:15:00 - ----------------------------------------
2025-02-14 23:15:00 - Script started.
2025-02-14 23:15:00 - Removing system information...
2025-02-14 23:15:05 - Removed system information successfully.
2025-02-14 23:15:09 - Extracting Lockdown Browser...
2025-02-14 23:15:49 - Installing Lockdown Browser...
2025-02-14 23:15:49 - Setup executed.
2025-02-14 23:15:53 - Removed existing LockDown Browser shortcut.
2025-02-14 23:15:53 - Registering URL protocol(s)...
2025-02-14 23:15:53 - Successfully set item property for URL protocol rldb.
2025-02-14 23:15:53 - Creating run script on desktop...
2025-02-14 23:15:54 - Removed existing LockDown Browser start menu shortcut.
2025-02-14 23:15:54 - Run script, shortcut, and start menu shortcut created.
2025-02-14 23:16:03 - Script completed.


=== setup.log ===

[InstallShield Silent]
Version=v7.00
File=Log File
[ResponseResult]
ResultCode=0
[Application]
Name=Respondus LockDown Browser 2
Version=2.00.0000
Company=Respondus
Lang=0409
```

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,68 @@
 $log_file_path = Join-Path -Path $PSScriptRoot -ChildPath "logs/Build.log"
-function Write-Log {
+function wtite_log {
     param (
```
omg, idk how I missed that one! Thanks!

Voidless7125 replied:
Fixed.

Voidless7125 commented on runtime_directory/sandbox_run.ps1:
```diff
@@ -97,7 +103,7 @@ function Register-URLProtocol {
         # I got the urls from installing LDB OEM and looking in the registry for :Lockdown Browser OEM and found all these HKCR keys.
         $urls = @("anst", "cllb", "ibz", "ielb", "jnld", "jzl", "ldb", "ldb1", "pcgs", "plb", "pstg", "rzi", "uwfb", "xmxg")
```
Sounds good.

Voidless7125 replied:
Done.

Voidless7125 commented:
I didn't mean to delete the comment; I was trying to hide it in vscode. Sorry about that.
Ready for review and merge on Monday.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -6,71 +6,118 @@
+}
+
+// Install all detour hooks
+void InstallDetourHooks()
```
This should be `static` (assuming that it doesn't break the build).

(Same thing for `UninstallDetourHooks` too)

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
The NBSP at the start of the line breaks the Markdown formatting.

Voidless7125 replied:
Oh, sorry about that. I'll fix that.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
Another NBSP here.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
`Browser` should probably be `_Browser_` for consistency.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
```diff
@@ -95,7 +95,13 @@ cd C:\Users\WDAGUtilityAccount\Desktop\runtime_directory\
+
+### Build issues
+
+If you have to build issues, please run `.\build.ps1 -Clean` to reset your workplace to a fresh start. It should build fine.
```
I'm not sure what you mean by `It should build fine.`?

Voidless7125 replied:
I ment as in it should fix everything, but it's jarring. I have removed it. (You may not see it yet, working on everything at once.)

gucci-on-fleek commented on Readme.md:
```diff
@@ -95,7 +95,13 @@ cd C:\Users\WDAGUtilityAccount\Desktop\runtime_directory\
+
+If you have to build issues, please run `.\build.ps1 -Clean` to reset your workplace to a fresh start. It should build fine.
+
+If you still have issues, run `.\build.ps1 -Logs` for logging into a zip file you can send us.
```
(not a zip file anymore)

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
```diff
@@ -107,20 +113,20 @@ This repo consists of simple tools cobbled together into a coherent package.
 The _Lockdown Browser_ detects a few BIOS-related registry keys in `HKLM:\HARDWARE\DESCRIPTION`. Therefore, `sandbox_run.ps1` deletes these keys/values.
 
-When the _Lockdown Browser_ detects that `VmComputeAgent.exe` is running, it realizes it is in a VM and refuses to launch. This program is part of the _Windows Sandbox_, and cannot be stopped without crashing the VM.  However, when the _Browser_ checks all the running programs, it also opens and examines each image file. If `sandbox_run.ps1` deletes the image file, the _Lockdown Browser_ acts as  ...
+When the _Lockdown Browser_ detects that `VmComputeAgent.exe` is running, it realizes it is in a VM and refuses to launch. This program is part of the _Windows Sandbox_, and cannot be stopped without crashing the VM.  However, when the _Browser_ checks all the running programs, it also opens and examines each image file. If `sandbox_run.ps1` deletes the image file, the _Lockdown Browser_ acts lik ...
```
Another NBSP that shouldn't be here.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
```diff
@@ -107,20 +113,20 @@ This repo consists of simple tools cobbled together into a coherent package.
 The _Lockdown Browser_ calls `GetSystemMetrics(SM_REMOTESESSION)` to determine if it runs in an RDP session. Since this function is in `user32.dll`, there aren’t any trivial ways to fix this. However, [_Microsoft Detours_](https://github.com/microsoft/Detours) allows you to intercept and replace any function in any `.dll`. A small hook (`GetSystemMetrics-Hook.cpp`) is used with `Detours` to inter ...
 
-Because this tool runs in the _Windows Sandbox_, no state is retained between sessions. Therefore, this tool provides a scripted installer for the _Lockdown Browser_. The _Lockdown Browser_’s installer is a little tricky to script, so the installation is a little hacky, but it works. And again, the _Sandbox_ is completely isolated from the rest of your system, so the _Lockdown Browser_ cannot cau ...
+Because this tool runs in the _Windows Sandbox_, no state is retained between sessions. Therefore, this tool provides a scripted installer for the _Lockdown Browser_. The _Lockdown Browser_’s installer is a little tricky to script, so the installation is a little hacky, but it works. And again, the _Sandbox_ is wholly isolated from the rest of your system, so the _Lockdown Browser_ cannot cause a ...
```
`wholly` feels weird here; leave it as `completely` I think.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
```diff
@@ -107,20 +113,20 @@ This repo consists of simple tools cobbled together into a coherent package.
 ## License
 
-All code is licensed under the [_Mozilla Public License_, version 2.0](https://www.mozilla.org/en-US/MPL/2.0/) or greater. The documentation is licensed under [CC-BY-SA, version 4.0](https://creativecommons.org/licenses/by-sa/4.0/legalcode) or greater, in addition to the MPL. The _Detours_ submodule has an MIT licence as detailed in `Detours/LICENSE.md`.
+All code is licensed under the [_Mozilla Public License_, version 2.0](https://www.mozilla.org/en-US/MPL/2.0/) or greater. The documentation is licensed under [CC-BY-SA, version 4.0](https://creativecommons.org/licenses/by-sa/4.0/legalcode) or greater, in addition to the MPL. The _Detours_ submodule has an MIT license as detailed in `Detours/LICENSE.md`.
```
Pfft, typical American, coming in here and telling a Canadian that he's spelling words wrong. I bet that you say “color” instead of “colour” too :)

Ok, joking aside, I do prefer `licence` over `license`, almost exclusively out of a sense of misguided patriotism.

Voidless7125 replied:
Even my mom laughed to this one, that was good! We don't how to "speel" do we?

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on Readme.md:
```diff
@@ -5,9 +5,9 @@
-<img src="docs/Example_1a.png" width="50%"/><img src="docs/Example_1b.png" width="50%" />
-<details><summary>View Demo Video</summary>
-<video src="https://user-images.githubusercontent.com/49086429/143725579-42cd9f47-4462-4a54-aa8a-8419bfdebb68.mp4" title="Demonstration video of the software."></video> <!-- Licence for Example_2.webm: MPL-2.0+ OR CC-BY-SA-4.0+ --></details>
+![Example screenshot](docs/Example_1.png)
+
+[View Demo Video](https://user-images.githubusercontent.com/49086429/143725579-42cd9f47-4462-4a54-aa8a-8419bfdebb68.mp4)
```
I think that the two full page images/videos is a bit much to have at the top of the Readme. Probably the best solution here would be to set `Example_1.png` as the first frame of `Example_2.webm` so that we'd get the best of both worlds.

I've opened #130 as a reminder for me to do this later, but if you know how to do this, then feel free to do it yourself.

Voidless7125 replied:
I could use a video editor to add that. (You would have to remux most likely)

gucci-on-fleek replied:
Probably shouldn't need to reencode it since most codecs support concatenating without reencoding. I'll play around with `ffmpeg`.

gucci-on-fleek replied:
Ok, I've merged both files into 1 _without_ reencoding the video. I've pushed the binary file changes in e2600277bb43ffd7c6692fc622474b16ed17dfd1, but I'll leave the `Readme.md` changes to you so that we don't get merge conflicts.

gucci-on-fleek commented on Readme.md:
Grammar/spelling fixes look really nice here, thanks! Also, what tool are you using for this? I can maybe add it to a GitHub action to run it against every future PR.

Voidless7125 replied:
Yeah, I am a bad speller so I use grammerly, (Yes you are now in a ad!) and it finds all the readme fixes I've done. I missed that it changed completely to wholly, I would of said *no*

gucci-on-fleek reviewed (changes requested):
This PR looks really good, just a couple tiny comments/changes. Thanks as always.

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,73 @@
+# If the log flag (-Logs) is passed, create a one logs file.
+if ($Logs) {
+    write_log "Logs being put in one file as requested by -logs flag"
+    $allLogsPath = Join-Path $PSScriptRoot "logs/all-logs.log"
```
I'm so sorry, i was tired XD

Voidless7125 replied:
Fixed.

Voidless7125 commented on build.ps1:
```diff
@@ -3,31 +3,73 @@
+if ($Logs) {
+    write_log "Logs being put in one file as requested by -logs flag"
+    $allLogsPath = Join-Path $PSScriptRoot "logs/all-logs.log"
+    if (Test-Path $allLogsPath) { Remove-Item $allLogsPath -Force }
```
👍

Voidless7125 replied:
Fixed.

Voidless7125 commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -6,71 +6,118 @@
 #include <windows.h>
 #include <detours.h>
```
You could, and it's not a internal header.

Voidless7125 commented on build.ps1:
```diff
@@ -67,51 +105,57 @@ function build_detours {
+    write_log "----------------------------------------"
+    write_log "Build script started"
+    write_log "Collecting system information..."
+    $sysInfo = systeminfo 2>&1 | Out-String
+    $regexFilter = 'Registered Owner|Time Zone|Wireless|Wi-Fi|Network Card|Bluetooth|DHCP|IP address|Connection Name|NIC|Status:|Realtek|Wintun|\b\d{1,3}(\.\d{1,3}){3}\b|fe80::'
+    $sysInfo -split "`n" | ForEach-Object {
+        if (($_ -ne "") -and ($_ -notmatch $regexFilter)) {
+            write_log $_.Trim()
+        }
```
Here's the build output now:
```log
    Directory: C:\Users\Dominic\Desktop\dvt\lockdown-browser


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         2/14/2025  11:26 PM                logs
2025-02-15 11:40:05 - ----------------------------------------
2025-02-15 11:40:05 - Build script started
2025-02-15 11:40:05 - Selected system information:
2025-02-15 11:40:05 - Architecture: 64-bit
2025-02-15 11:40:05 - Windows Version: Microsoft Windows 11 Pro
2025-02-15 11:40:05 - Windows Build: 26100
2025-02-15 11:40:05 - Windows Edition: Pro
2025-02-15 11:40:10 - Antivirus Real-Time Protection: True
2025-02-15 11:40:10 - No VPN connections found.
2025-02-15 11:40:10 - Visual Studio Instances:
2025-02-15 11:40:10 -  - C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools
2025-02-15 11:40:10 -  - C:\Program Files\Microsoft Visual Studio\2022\Community
2025-02-15 11:40:10 - Repository version: v1.4.0_RC2-40-g6bd15d4
2025-02-15 11:40:11 - Submodule status:  b2bf32a657be1114b173ad28d2d65463ca4c466f Detours (v4.0.1-112-gb2bf32a)
2025-02-15 11:40:11 - Complete Visual Studio Setup Instances:
2025-02-15 11:40:11 -

InstanceId          : 2446a6bd
DisplayName         : Visual Studio Build Tools 2022                                                                    InstallationVersion : 17.13.35806.99                                                                                    InstallationPath    : C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools                                    InstallDate         : 2/11/2025 10:41:33 AM                                                                                                                                                                                                     InstanceId          : 77165d51
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.13.35806.99
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Community
InstallDate         : 2/11/2025 9:33:48 AM




2025-02-15 11:40:11 - Initializing Visual Studio environment
WARNING: The version '2.2.16' of module 'VSSetup' is currently in use. Retry the operation after closing the
applications. (Note not from script, I'll fix this now that I need elavation to check if the windows sandbox is installed.)
2025-02-15 11:40:22 - Visual Studio environment initialized
2025-02-15 11:40:22 - Building Detours

Microsoft (R) Program Maintenance Utility Version 14.43.34808.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Microsoft (R) Program Maintenance Utility Version 14.43.34808.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Microsoft (R) Program Maintenance Utility Version 14.43.34808.0
Copyright (C) Microsoft Corporation.  All rights reserved.

2025-02-15 11:40:23 - Detours built
2025-02-15 11:40:23 - Building hook
d-----         2/15/2025  11:40 AM                build
Microsoft (R) C/C++ Optimizing Compiler Version 19.43.34808 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

GetSystemMetrics-Hook.cpp
Microsoft (R) Incremental Linker Version 14.43.34808.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/dll
/implib:GetSystemMetrics-Hook.lib
/out:GetSystemMetrics-Hook.dll
/nodefaultlib:oldnames.lib
/export:DetourFinishHelperProcess,@1,NONAME
/export:GetSystemMetrics
../Detours\lib.X86\detours.lib
../Detours\lib.X86\syelog.lib
user32.lib
GetSystemMetrics-Hook.obj
   Creating library GetSystemMetrics-Hook.lib and object GetSystemMetrics-Hook.exp
2025-02-15 11:40:24 - Hook built
2025-02-15 11:40:24 - Building sandbox configuration
2025-02-15 11:40:24 - Sandbox configuration built
2025-02-15 11:40:24 - Copying files to runtime directory
2025-02-15 11:40:25 - Files copied to runtime directory
Do you want to add the runtime directory ('C:\Users\Dominic\Desktop\dvt\lockdown-browser\runtime_directory') as an exclusion for Windows Defender? (Y/N): Y
2025-02-15 11:40:29 - Runtime directory added as exclusion for Windows Defender.
2025-02-15 11:40:29 - Build script completed
```

Voidless7125 commented on build.ps1:
```diff
@@ -67,51 +105,57 @@ function build_detours {
     (Get-Content ./src/Sandbox-with-Microphone-Camera.xml) -replace '{{HOST_FOLDER}}', $host_folder_path -replace '{{LOG_FOLDER}}', $log_folder_path | Set-Content ./build/Sandbox-with-Microphone-Camera.wsb
-    
-    Write-Log "Sandbox configuration built"
+    write_log "Sandbox configuration built"
```
Fixed.

Voidless7125 commented on runtime_directory/sandbox_run.ps1:
```diff
@@ -109,10 +115,18 @@ function Register-URLProtocol {
 function New-RunLockdownBrowserScript {
     Write-Log "Creating run script on desktop..."
+
+    $browserIcon = if ($is_oem) {
```
Fixed.

Voidless7125 commented on runtime_directory/sandbox_run.ps1:
```diff
@@ -131,23 +145,47 @@ Set-Location C:\Users\WDAGUtilityAccount\Desktop\runtime_directory\
     $result = [System.Windows.Forms.MessageBox]::Show("Do you want to test launch Lockdown Browser to ensure that there are no errors? (Highly recommended).", "Test LockDown Browser", [System.Windows.Forms.MessageBoxButtons]::YesNo, [System.Windows.Forms.MessageBoxIcon]::Question)
     if ($result -eq [System.Windows.Forms.DialogResult]::Yes) {
         if ($is_oem) {
-            # URL is from https://github.com/gucci-on-fleek/lockdown-browser/issues/43.
```
Fixed.

Voidless7125 commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -6,71 +6,118 @@
 
-BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+// Custom hook functions for TerminateProcess and ExitProcess
+BOOL WINAPI MyTerminateProcess(HANDLE hProcess, UINT uExitCode)
```
Fixed.

Voidless7125 commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -6,71 +6,118 @@
+    {
+        DetourTransactionAbort();
+        return;
+    }
```
Fixed.

Voidless7125 commented on .github/ISSUE_TEMPLATE/bug-report.yml:
```diff
@@ -113,7 +116,7 @@ body:
       label: Other
       description: Add any other relevant information here. (Optional)
-
+      
```
Fixed.

Voidless7125 commented on .github/ISSUE_TEMPLATE/bug-report.yml:
```diff
@@ -27,6 +27,8 @@ body:
             If you get a "the setup configuration provider is not registered" error message, have you [installed the "Visual Studio C++ Tools"](https://github.com/gucci-on-fleek/lockdown-browser?tab=readme-ov-file#system-requirements)?
           required: true
+        - label: > 
+            You have tried reseting your build with `./build.ps1 -Clean`
```
Done.

gucci-on-fleek commented on Readme.md:
```diff
@@ -1,6 +1,6 @@
 <!-- Lockdown Browser in Windows Sandbox
      https://github.com/gucci-on-fleek/lockdown-browser
-     SPDX-License-Identifier: MPL-2.0+ OR CC-BY-SA-4.0+
+     SPDX-licence-Identifier: MPL-2.0+ OR CC-BY-SA-4.0+
```
Ah, but this one should stay as `License` since it's a standardized keyword, and not an English word (even though it happens to correspond exactly with an English word).

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on build.ps1:
```diff
@@ -1,21 +1,25 @@
+    Write-Host "Restarting as Administrator..."
+    Start-Process powershell -ArgumentList "-NoProfile", "-ExecutionPolicy", "Bypass", "-File", $PSCommandPath -Verb RunAs
+    exit
+}
```
Hmm, why do we need to be running as an Administrator? I don't think that this should be necessary.

Voidless7125 replied:
Need it to check the status of windows sandbox.

gucci-on-fleek replied:
Hmm, okay, so let's just remove the _Windows Sandbox_ status testing stuff then.

Voidless7125 replied:
I figured out a way to do it without admin, all good.

gucci-on-fleek commented on build.ps1:
```diff
@@ -105,57 +108,128 @@ function build_detours {
     mkdir './build' -Force
     Push-Location build
-    cl '/EHsc' '/LD' '/Fe:GetSystemMetrics-Hook.dll' '../src/GetSystemMetrics-Hook.cpp' '/I../Detours/include' '/link' '/nodefaultlib:oldnames.lib' '/export:DetourFinishHelperProcess,@1,NONAME' '/export:GetSystemMetrics' '../Detours\lib.X86\detours.lib' '../Detours\lib.X86\syelog.lib' 'user32.lib'  # Most of these are pretty standard VS C++ compiler options, but of note is "/export:DetourFinishHe ...
+    cl '/EHsc' `
```
Much better on multiple lines, thanks. But maybe add the comment back in (or a different comment mentioning the `/export:DetourFinishHelperProcess,@1,NONAME` stuff.

Voidless7125 replied:
Fixed.

gucci-on-fleek commented on build.ps1:
```diff
@@ -105,57 +108,128 @@ function build_detours {
-    $sysInfo -split "`n" | ForEach-Object {
-        if (($_ -ne "") -and ($_ -notmatch $regexFilter)) {
-            write_log $_.Trim()
+function Add-AVExclusion {
+    $runtimePath = Join-Path $PSScriptRoot "runtime_directory"
+    $response = Read-Host "Do you want to add the runtime directory ('$runtimePath') as an exclusion for Windows Defender? (Y/N)"
+    if ($response -match '^[yY]$') {
+        Add-MpPreference -ExclusionPath $runtimePath
+        Write-Log "Runtime directory added as exclusion for Windows Defender."
+    }
+    else {
+        Write-Log "User skipped adding runtime directory to Windows Defender exclusions."
+    }
+}
```
Hmm, this is definitely useful, but I don't think that we should do this. The point of this script is to be more trustworthy than Respondus, but this is rather sketchy (albeit justified).

Voidless7125 replied:
Np.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -38,28 +38,29 @@ VOID WINAPI MyExitProcess(UINT uExitCode)
 
-    // Hook GetSystemMetrics
-    if (DetourAttach(reinterpret_cast<PVOID *>(&Original_GetSystemMetrics), Hooked_GetSystemMetrics) != NO_ERROR)
+    struct HookPair
```
I think that it would be better to declare the struct type at the top-level instead of inside of a function.

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -38,28 +38,29 @@ VOID WINAPI MyExitProcess(UINT uExitCode)
+        PVOID pHook;
+    };
+
+    HookPair hooks[] = {
+        {reinterpret_cast<PVOID *>(&Original_GetSystemMetrics), reinterpret_cast<PVOID>(Hooked_GetSystemMetrics)},
+        {reinterpret_cast<PVOID *>(&Original_TerminateProcess), reinterpret_cast<PVOID>(Hooked_TerminateProcess)},
+        {reinterpret_cast<PVOID *>(&Original_ExitProcess), reinterpret_cast<PVOID>(MyExitProcess)}};
```
This is used here and in `UninstallDetourHooks`, so could we have this be module global (aka `static const` at the top of the file)?

gucci-on-fleek commented on src/GetSystemMetrics-Hook.cpp:
```diff
@@ -38,28 +38,29 @@ VOID WINAPI MyExitProcess(UINT uExitCode)
+            DetourTransactionAbort();
+            return;
+        }
     }
```
I have very little C++ experience, so I really have no idea here, but do iterators dynamically allocate any memory? Because it's dangerous to use `malloc` and similar inside of `DllMain`. I'd probably guess that this is safe though.

Voidless7125 replied:
Reverted changes as per convo.

gucci-on-fleek reviewed (commented):
Looks good, just a couple tiny changes.

Voidless7125 commented:
### Documentation updates:
* [`Readme.md`](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L3-R8): Updated various sections for clarity, corrected typos, and improved formatting. Notable changes include updating the system requirements, adjusting instructions for building and running the project, and refining the disclaimer and support sections. [[1]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L3-R8) [[2]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L39-R52) [[3]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L66-R64) [[4]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L81-R87) [[5]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L98-R102) [[6]](diffhunk://#diff-1550ec65ac92f65817fc28928dfef526912b5f52356ff43651369bae92f56031L110-R130)

### Issue template improvements:
* [`.github/ISSUE_TEMPLATE/bug-report.yml`](diffhunk://#diff-7c49a365610d935ce439e3f241cd3f5d1df89a95d1a674d11ebdb1e518c7335cR30-R31): Added new steps for troubleshooting, such as rebuilding from scratch and providing log file paths. Updated placeholders for Lockdown Browser and Windows versions. [[1]](diffhunk://#diff-7c49a365610d935ce439e3f241cd3f5d1df89a95d1a674d11ebdb1e518c7335cR30-R31) [[2]](diffhunk://#diff-7c49a365610d935ce439e3f241cd3f5d1df89a95d1a674d11ebdb1e518c7335cL59-R63) [[3]](diffhunk://#diff-7c49a365610d935ce439e3f241cd3f5d1df89a95d1a674d11ebdb1e518c7335cL82-R85) [[4]](diffhunk://#diff-7c49a365610d935ce439e3f241cd3f5d1df89a95d1a674d11ebdb1e518c7335cL95-R98)

### Pull request template update:
* [`.github/pull_request_template.md`](diffhunk://#diff-b2496e80299b8c3150b1944450bd81c622e04e13d15c411d291db0927d75fd6bL16): Minor formatting adjustment to improve readability.

### Build script enhancements:
* [`build.ps1`](diffhunk://#diff-3258bc4162690adead72c70b672093e68c92e86a4628148f0015d61738f21b5aR6-L59): Introduced new launch parameters (`-Clean` and `-Logs`) to manage build cleanup and logging. Added checks for Git installation and improved Visual Studio environment initialization. Fixed Visual Studio instances bug, added system info for diag purposes.

### Sandbox script enhancements
* [`sandbox_run.ps1`](diffhunk://##diff-2bae01233a6adc333d3582c1de57f9a97e5d1dbbdcdcf2f40bc4d3dd9310a537):
Pop-up warnings when something goes wrong. Faster desktop icon loading (By 20-23 seconds!), Full support for OEM browsers, Start menu shortcut changed.

Voidless7125 commented:
Thanks for that, I forgot *again*

gucci-on-fleek reviewed (approved):
Perfect, thanks! I'll merge this shortly.

(There are still a couple of tiny changes that I want to make before releasing v1.4, so please wait before tagging that. I'll open PRs for those later tonight.)
